src/cim/components/ClusterDeployment/ClusterDeploymentHostsSelection.tsx

```diff
@@ -0,0 +1,96 @@
+import React from 'react';
+import { useFormikContext } from 'formik';
+import { Grid, GridItem, TextContent } from '@patternfly/react-core';
+import { ClusterWizardStepHeader, SwitchField } from '../../../common';
+import {
+  ClusterDeploymentHostsSelectionProps,
+  ClusterDeploymentHostsSelectionValues,
+} from './types';
+import ClusterDeploymentHostsSelectionBasic from './ClusterDeploymentHostsSelectionBasic';
+import ClusterDeploymentHostsSelectionAdvanced from './ClusterDeploymentHostsSelectionAdvanced';
+import ConfirmationModal from '../../../common/components/ui/ConfirmationModal';
+
+const ClusterDeploymentHostsSelection: React.FC<ClusterDeploymentHostsSelectionProps> = ({
+  agentLocations,
+  matchingAgents,
+  onAgentSelectorChange,
+  // allAgentsCount,
+  usedAgentLabels,
+  hostActions,
+}) => {
+  const { setFieldValue, values } = useFormikContext<ClusterDeploymentHostsSelectionValues>();
+  const [isConfirmAutoSelectHostsSwitchOpen, setConfirmAutoSelectHostsSwitch] = React.useState(
+    false,
+  );
+  const { autoSelectHosts, isSNOCluster } = values;
+
+  const onChangeCustomOverride =
+    !values.autoSelectHosts && (values.agentLabels.length > 0 || values.selectedHostIds.length > 0)
+      ? () => {
+          setConfirmAutoSelectHostsSwitch(true);
+        }
+      : undefined;
+
+  const onAutoSelectConfirmed = () => {
+    setFieldValue('autoSelectHosts', !values.autoSelectHosts);
+    setFieldValue('agentLabels', []);
+    setFieldValue('selectedHostIds', []);
+    setConfirmAutoSelectHostsSwitch(false);
+  };
+
+  return (
+    <Grid hasGutter>
+      <GridItem>
+        <ClusterWizardStepHeader>Hosts selection</ClusterWizardStepHeader>
+        <TextContent>
+          {isSNOCluster
+            ? 'Exactly 1 host is required, capable of functioning both as control plane (master) and worker node.'
+            : 'At least 3 hosts are required, capable of functioning as control plane (master) nodes.'}
+        </TextContent>
+      </GridItem>
+
+      <GridItem>
+        <SwitchField
```

Comment:
How about instead of extending the API of the generic `SwitchField` define a custom `AutoSelectHostsSwitchField` which will encapsulate all custom logic including the confirmation modal?
I think we should not be afraid of defining custom Formik fields for cases where inputs have specific behaviour.